### ISSUE TITLE
`adapt_structure` for `AbstractArray` boundary conditions invokes `parent`

### ISSUE THREAD

**Issue by glwagner** [CliMA/Oceananigans.jl#3286]
This prevents `Field` and `OffsetArray` from being used as boundary conditions (correctly) --- if users try to use `Field`, they get different behavior on CPU vs GPU.

I'm not sure the motivation for this choice. 

https://github.com/CliMA/Oceananigans.jl/blob/e16cdc6cfb67703df9e29368017868331f68b1c0/src/BoundaryConditions/boundary_condition.jl#L69-L70

**Comment by glwagner** [comment]
It looks like it dates back to ancient times practically before the invention of writing

https://github.com/CliMA/Oceananigans.jl/commit/7f4b78ef3cbca33d1f23c45e978d223a82566623#diff-ef43fe13b0bf0810d1bd93a5cd3264c3be46ba792948fad6a924a62b281e3286

I'm amazed nobody has caught this bug before.

**Comment by navidcy** [comment]
It may be the answer to several conundrums. Or simply not many use fields for BCs?

**Comment by glwagner** [comment]
@simone-silvestri may have tried. Not sure. @yuchenma23?

**Comment by yuchenma23** [comment]
Hi Greg, I was trying field as boundary condition but with time_interpolated_array Simone defined, which is used to interpolate field as a function of time. Essentially it's still a field boundary condition. It never worked on GPU with the bug that is related with some mismatch between types things. But it did work on CPU. 

I then got to work on other stuff and never went back to check the code. Perhaps the bug you fixed could explain the bug I faced.
